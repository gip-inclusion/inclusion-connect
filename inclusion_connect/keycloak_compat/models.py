from django.db import models


class JWTHashSecret(models.Model):
    realm_id = models.TextField(primary_key=True)
    # urlsafe_base64_encoded.
    secret = models.TextField()
