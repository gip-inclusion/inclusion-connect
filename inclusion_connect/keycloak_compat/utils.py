import re


namespace_pattern = re.compile(r"^keycloak_compat_(?P<realm>.+)$")


def realm_from_request(request):
    namespace = request.resolver_match.namespace
    match = namespace_pattern.match(namespace)
    return match.group("realm")
