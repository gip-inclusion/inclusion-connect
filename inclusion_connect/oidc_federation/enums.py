from django.db import models


class Federation(models.TextChoices):
    PEAMA = "peama", "France Travail"
