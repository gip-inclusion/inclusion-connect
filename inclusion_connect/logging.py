import logging.handlers
import threading

from elasticsearch import Elasticsearch
from elasticsearch.helpers import bulk
from pythonjsonlogger import jsonlogger

from inclusion_connect.utils.oidc import oidc_params


def log_data(request, next_url=None):
    log_data = {"ip_address": request.META["REMOTE_ADDR"]}
    params = oidc_params(request, next_url)
    try:
        log_data["application"] = params["client_id"]
    except KeyError:
        pass
    return log_data


class JsonFormatter(jsonlogger.JsonFormatter):
    def parse(self):
        # Remove the empty key "message".
        return []

    def add_fields(self, log_record, record, message_dict):
        super().add_fields(log_record, record, message_dict)
        for record_attr in ["name", "levelname"]:
            log_record[record_attr] = getattr(record, record_attr)


class ElasticSearchHandler(logging.handlers.BufferingHandler):
    timer = None

    def __init__(  # noqa: PLR0913 Too many arguments to function call.
        self,
        *,
        capacity,
        send_after_inactive_for_secs=2.0,
        host,
        index_name,
        es_timeout_secs=5,
    ):
        """
        Inspired from https://github.com/cmanaha/python-elasticsearch-logger/blob/master/cmreslogging/handlers.py.

        :param int capacity: Max number of log records before sending to ElasticSearch.
        :param float send_after_inactive_for_secs: Send to ElasticSearch if no records
            are emitted in send_after_inactive_for_secs seconds.
        """
        super().__init__(capacity)
        self.send_after_inactive_for_secs = send_after_inactive_for_secs
        self.es_client = Elasticsearch(host, http_compress=True, request_timeout=es_timeout_secs, max_retries=10)
        self.index_name = index_name

    def emit(self, record):
        if self.timer and self.timer.is_alive():
            self.timer.cancel()
        formatted_record = self.format(record)
        super().emit(formatted_record)
        self.timer = threading.Timer(self.send_after_inactive_for_secs, self.flush)
        self.timer.start()

    def send_to_elastic(self, log_buffer):
        actions = ({"_source": log} for log in log_buffer)
        bulk(client=self.es_client, actions=actions, index=self.index_name, stats_only=True)

    def flush(self):
        with self.lock:
            log_buffer = self.buffer
            self.buffer = []
        if log_buffer:
            t = threading.Thread(target=self.send_to_elastic, args=(log_buffer,), daemon=False)
            t.start()
