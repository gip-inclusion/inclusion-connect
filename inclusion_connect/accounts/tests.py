from urllib.parse import quote

from django.contrib import messages
from django.contrib.auth import get_user
from django.core import mail
from django.urls import reverse
from django.utils.encoding import force_bytes
from django.utils.http import urlsafe_base64_encode
from pytest_django.asserts import assertContains, assertRedirects, assertTemplateUsed

from inclusion_connect.accounts.views import PasswordResetView
from inclusion_connect.users.factories import DEFAULT_PASSWORD, UserFactory
from inclusion_connect.users.models import User
from inclusion_connect.utils.urls import add_url_params


def test_login(client):
    redirect_url = reverse("oidc_overrides:logout")
    url = add_url_params(reverse("accounts:login"), {"next": redirect_url})
    user = UserFactory()
    assert not get_user(client).is_authenticated

    response = client.get(url)
    assertContains(response, "Connexion")
    assertContains(response, "Adresse e-mail")  # Ask for email, not username
    assertContains(response, reverse("accounts:registration"))  # Link to registration page

    response = client.post(url, data={"email": user.email, "password": DEFAULT_PASSWORD})
    assertRedirects(response, redirect_url, fetch_redirect_response=False)
    assert get_user(client).is_authenticated


def test_login_failed_bad_email_or_password(client):
    url = add_url_params(reverse("accounts:login"), {"next": "anything"})
    user = UserFactory()
    assert not get_user(client).is_authenticated

    response = client.post(url, data={"email": user.email, "password": "toto"})
    assertTemplateUsed("login.html")
    assertContains(response, "Adresse e-mail ou mot de passe invalide.")
    assert not get_user(client).is_authenticated

    response = client.post(url, data={"email": "wrong@email.com", "password": DEFAULT_PASSWORD})
    assertTemplateUsed("login.html")
    assertContains(response, "Adresse e-mail ou mot de passe invalide.")
    assert not get_user(client).is_authenticated

    # If user is inactive
    user.is_active = False
    user.save()
    response = client.post(url, data={"email": user.email, "password": DEFAULT_PASSWORD})
    assertTemplateUsed("login.html")
    assertContains(response, "Adresse e-mail ou mot de passe invalide.")
    assert not get_user(client).is_authenticated


def test_user_creation(client):
    redirect_url = reverse("oidc_overrides:logout")
    url = add_url_params(reverse("accounts:registration"), {"next": redirect_url})
    user = UserFactory.build()
    assert not get_user(client).is_authenticated

    response = client.get(url)
    assertContains(response, "Créer un compte")
    assertContains(response, reverse("accounts:login"))  # Link to login page
    assertContains(response, "CGU_20230302.pdf")
    assertContains(response, quote("Politique de confidentialité_20230302.pdf"))

    response = client.post(
        url,
        data={
            "email": user.email,
            "first_name": user.first_name,
            "last_name": user.last_name,
            "password1": DEFAULT_PASSWORD,
            "password2": DEFAULT_PASSWORD,
            "terms_accepted": True,
        },
    )
    assertRedirects(response, redirect_url, fetch_redirect_response=False)
    assert get_user(client).is_authenticated
    user = User.objects.get(email=user.email)  # Previous instance was a built factory, so refresh_from_db won't work
    assert user.terms_accepted_at == user.date_joined


def test_user_creation_terms_are_required(client):
    redirect_url = reverse("oidc_overrides:logout")
    url = add_url_params(reverse("accounts:registration"), {"next": redirect_url})
    user = UserFactory.build()
    assert not get_user(client).is_authenticated

    response = client.post(
        url,
        data={
            "email": user.email,
            "first_name": user.first_name,
            "last_name": user.last_name,
            "password1": DEFAULT_PASSWORD,
            "password2": DEFAULT_PASSWORD,
        },
    )
    assertTemplateUsed("registration.html")
    assert "terms_accepted" in response.context["form"].errors


# TODO: Test next url propagation in all the registration process (creation + email validation)


def test_password_reset(client):
    user = UserFactory()

    redirect_url = reverse("oidc_overrides:logout")
    url = add_url_params(reverse("accounts:login"), {"next": redirect_url})
    response = client.get(url)
    password_reset_url = reverse("accounts:password_reset")
    assertContains(response, password_reset_url)

    response = client.get(password_reset_url)
    assertTemplateUsed(response, "password_reset.html")

    response = client.post(password_reset_url, data={"email": user.email})
    assertRedirects(response, reverse("accounts:login"))
    assert list(messages.get_messages(response.wsgi_request)) == [
        messages.storage.base.Message(
            messages.SUCCESS,
            "Si un compte existe avec cette adresse e-mail, "
            "vous recevrez un e-mail contenant des instructions pour réinitialiser votre mot de passe.",
        ),
    ]

    # Check sent email
    assert len(mail.outbox) == 1
    uid = urlsafe_base64_encode(force_bytes(user.pk))
    token = PasswordResetView.token_generator.make_token(user)
    password_reset_url = reverse("accounts:password_reset_confirm", args=(uid, token))
    assert password_reset_url in mail.outbox[0].body

    # Change password
    password = "toto"
    response = client.get(password_reset_url)  # retrieve the modified url
    response = client.post(response.url, data={"new_password1": password, "new_password2": password})
    # FIXME: skip redirection to login, go to next url !
    assertRedirects(response, redirect_url, fetch_redirect_response=False)
    assert get_user(client).is_authenticated
