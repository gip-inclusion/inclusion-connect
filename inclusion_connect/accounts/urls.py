from django.urls import path, re_path

from inclusion_connect.accounts import views


app_name = "accounts"

urlpatterns = [
    re_path(r"^login/$", views.LoginView.as_view(), name="login"),
    re_path(r"^register/$", views.RegisterView.as_view(), name="register"),
    re_path(r"^activate/$", views.ActivateAccountView.as_view(), name="activate"),
    re_path(r"^password_reset/$", views.PasswordResetView.as_view(), name="password_reset"),
    path("reset/<uidb64>/<token>/", views.PasswordResetConfirmView.as_view(), name="password_reset_confirm"),
    re_path(r"^accept-terms/$", views.AcceptTermsView.as_view(), name="accept_terms"),
    re_path(r"^my-account/$", views.EditUserInfoView.as_view(), name="edit_user_info"),
    re_path(r"^change-password/$", views.PasswordChangeView.as_view(), name="change_password"),
]
