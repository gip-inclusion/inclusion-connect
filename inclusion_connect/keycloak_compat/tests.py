import uuid

import jwt
import pytest
from django.contrib.auth import get_user
from django.contrib.auth.hashers import PBKDF2PasswordHasher
from django.urls import reverse
from oauth2_provider.models import get_access_token_model, get_id_token_model, get_refresh_token_model
from pytest_django.asserts import assertRedirects

from inclusion_connect.keycloak_compat.hashers import KeycloakPasswordHasher
from inclusion_connect.oidc_overrides.factories import DEFAULT_CLIENT_SECRET, ApplicationFactory, default_client_secret
from inclusion_connect.users.factories import UserFactory
from inclusion_connect.utils.urls import add_url_params, get_url_params


@pytest.mark.parametrize("realm", ["local", "Review_apps", "Demo", "inclusion-connect"])
def test_login(client, realm):
    application = ApplicationFactory()
    user = UserFactory()

    # Test AUTH endpoint when not authenticated
    auth_url = reverse(f"keycloak_compat_{realm}:authorize")
    auth_params = {
        "response_type": "code",
        "client_id": application.client_id,
        "redirect_uri": "http://localhost/callback",
        "scope": "openid profile email",
        "state": "state",
        "nonce": "nonce",
    }
    auth_complete_url = add_url_params(auth_url, auth_params)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:login"))
    assert client.session["next_url"] == auth_complete_url

    # Test AUTH endpoint when not authenticated and with bad params
    bad_auth_params = auth_params.copy()
    bad_auth_params["client_id"] = "toto"
    bad_auth_complete_url = add_url_params(auth_url, bad_auth_params)
    response = client.get(bad_auth_complete_url)
    # FIXME update the template
    assert response.status_code == 400

    # Test AUTH endpoint when authenticated
    client.force_login(user)
    response = client.get(auth_complete_url)
    assert response.status_code == 302
    assert response.url.startswith(auth_params["redirect_uri"])
    auth_response_params = get_url_params(response.url)

    # Test TOKEN endpoint
    # FIXME it's recommanded to use basic auth here, maybe update our documentation ?
    token_data = {
        "client_id": application.client_id,
        "client_secret": DEFAULT_CLIENT_SECRET,
        "code": auth_response_params["code"],
        "grant_type": "authorization_code",
        "redirect_uri": "http://localhost/callback",
    }
    response = client.post(
        reverse(f"keycloak_compat_{realm}:token"),
        data=token_data,
    )
    token_json = response.json()
    id_token = token_json["id_token"]
    decoded_id_token = jwt.decode(
        id_token,
        key=default_client_secret(),
        algorithms=["HS256"],
        audience=application.client_id,
    )
    assert decoded_id_token["nonce"] == auth_params["nonce"]
    assert decoded_id_token["sub"] == str(user.pk)
    assert uuid.UUID(decoded_id_token["sub"]), "Sub should be an uuid"
    assert decoded_id_token["given_name"] == user.first_name
    assert decoded_id_token["family_name"] == user.last_name
    assert decoded_id_token["email"] == user.email

    # Test USER INFO endpoint
    response = client.get(
        reverse(f"keycloak_compat_{realm}:user-info"),
        HTTP_AUTHORIZATION=f"Bearer {token_json['access_token']}",
    )
    assert response.json() == {
        "sub": str(user.pk),
        "given_name": user.first_name,
        "family_name": user.last_name,
        "email": user.email,
    }

    # Test LOGOUT endpoint
    assert get_user(client).is_authenticated
    logout_params = {"id_token_hint": id_token}
    response = client.get(add_url_params(reverse(f"keycloak_compat_{realm}:logout"), logout_params))
    assert not get_user(client).is_authenticated
    assert get_id_token_model().objects.count() == 0
    assert get_access_token_model().objects.count() == 0
    assert get_refresh_token_model().objects.get().revoked is not None


@pytest.mark.parametrize("realm", ["local", "Review_apps", "Demo", "inclusion-connect"])
def test_registration(client, realm):
    application = ApplicationFactory()
    user = UserFactory()

    # Test REGISTRATIONS endpoint when not authenticated
    auth_url = reverse(f"keycloak_compat_{realm}:registrations")
    auth_params = {
        "response_type": "code",
        "client_id": application.client_id,
        "redirect_uri": "http://localhost/callback",
        "scope": "openid profile email",
        "state": "state",
        "nonce": "nonce",
    }
    auth_complete_url = add_url_params(auth_url, auth_params)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:registration"))
    assert client.session["next_url"] == auth_complete_url

    # Test REGISTRATIONS endpoint when not authenticated and with bad oidc params
    bad_auth_params = auth_params.copy()
    bad_auth_params["client_id"] = "toto"
    bad_auth_complete_url = add_url_params(auth_url, bad_auth_params)
    response = client.get(bad_auth_complete_url)
    # FIXME update the template
    assert response.status_code == 400

    # Test REGISTRATIONS endpoint when not authenticated with activation params
    activation_params = auth_params | {
        "login_hint": "email",
        "firstname": "John",
        "lastname": "Doe",
    }
    auth_complete_url = add_url_params(auth_url, activation_params)
    response = client.get(auth_complete_url, follow=True)
    assertRedirects(response, reverse("accounts:activation"))
    assert client.session["next_url"] == auth_complete_url

    # Test REGISTRATIONS endpoint when authenticated
    client.force_login(user)
    response = client.get(auth_complete_url)
    assert response.status_code == 302
    assert response.url.startswith(auth_params["redirect_uri"])
    auth_response_params = get_url_params(response.url)

    # Test TOKEN endpoint
    # FIXME it's recommanded to use basic auth here, maybe update our documentation ?
    token_data = {
        "client_id": application.client_id,
        "client_secret": DEFAULT_CLIENT_SECRET,
        "code": auth_response_params["code"],
        "grant_type": "authorization_code",
        "redirect_uri": "http://localhost/callback",
    }
    response = client.post(
        reverse(f"keycloak_compat_{realm}:token"),
        data=token_data,
    )
    token_json = response.json()
    id_token = token_json["id_token"]
    decoded_id_token = jwt.decode(
        id_token,
        key=default_client_secret(),
        algorithms=["HS256"],
        audience=application.client_id,
    )
    assert decoded_id_token["nonce"] == auth_params["nonce"]
    assert decoded_id_token["sub"] == str(user.pk)
    assert uuid.UUID(decoded_id_token["sub"]), "Sub should be an uuid"
    assert decoded_id_token["given_name"] == user.first_name
    assert decoded_id_token["family_name"] == user.last_name
    assert decoded_id_token["email"] == user.email

    # Test USER INFO endpoint
    response = client.get(
        reverse(f"keycloak_compat_{realm}:user-info"),
        HTTP_AUTHORIZATION=f"Bearer {token_json['access_token']}",
    )
    assert response.json() == {
        "sub": str(user.pk),
        "given_name": user.first_name,
        "family_name": user.last_name,
        "email": user.email,
    }

    # Test LOGOUT endpoint
    assert get_user(client).is_authenticated
    logout_params = {"id_token_hint": id_token}
    response = client.get(add_url_params(reverse(f"keycloak_compat_{realm}:logout"), logout_params))
    assert not get_user(client).is_authenticated
    assert get_id_token_model().objects.count() == 0
    assert get_access_token_model().objects.count() == 0
    assert get_refresh_token_model().objects.get().revoked is not None


def test_password_hasher(client):
    password = "RdaRfqP7Y89vy2"
    hashed_password = "$".join(
        [
            "keycloak-pbkdf2-sha256",
            "27500",
            "Td6XuopYK6JNfUnIlqYMOQ==",
            "ZXVC08Hf4jBOoYzVoNWYjQijsMC2oc/OUa9LciiIJ/1XHPF/qPiY1DqwLLDN2hYFmf/1kApkveD8/Pr7GVqjgw==",
        ]
    )
    user = UserFactory(password=hashed_password)
    assert user.password == hashed_password

    assert KeycloakPasswordHasher().verify(password=password, encoded=hashed_password)

    client.login(email=user.email, password=password)

    user.refresh_from_db()
    assert user.password != hashed_password
    assert PBKDF2PasswordHasher().verify(password, encoded=user.password)
