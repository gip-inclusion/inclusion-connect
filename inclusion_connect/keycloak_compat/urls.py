from django.urls import re_path
from oauth2_provider import views as oauth2_views

from ..accounts.views import EditUserInfoView
from ..oidc_overrides.views import AuthorizationView, LogoutView, RegistrationView
from . import views


app_name = "keycloak_compat"


urlpatterns = [
    re_path(
        r"^\.well-known/openid-configuration/$",
        oauth2_views.ConnectDiscoveryInfoView.as_view(),
        name="oidc-connect-discovery-info",
    ),
    re_path(r"^protocol/openid-connect/userinfo$", oauth2_views.UserInfoView.as_view(), name="user-info"),
    re_path(r"^protocol/openid-connect/auth$", AuthorizationView.as_view(), name="authorize"),
    re_path(r"^protocol/openid-connect/registrations$", RegistrationView.as_view(), name="registrations"),
    re_path(r"^protocol/openid-connect/token$", oauth2_views.TokenView.as_view(), name="token"),
    re_path(r"^protocol/openid-connect/logout$", LogoutView.as_view(), name="logout"),
    re_path(r"^account$", EditUserInfoView.as_view(), name="edit_user_info"),
    re_path(r"^login-actions/action-token$", views.ActionToken.as_view(), name="action-token"),
]
