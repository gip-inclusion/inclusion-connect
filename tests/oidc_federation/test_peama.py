import logging
from types import SimpleNamespace

import jwt
from django.conf import settings
from django.contrib.auth import get_user
from django.test import override_settings
from django.urls import reverse
from jwcrypto import jwk
from pytest_django.asserts import assertContains, assertRedirects

from inclusion_connect.oidc_federation.enums import Federation
from inclusion_connect.users.models import User
from inclusion_connect.utils.urls import get_url_params
from tests.asserts import assertRecords
from tests.users.factories import DEFAULT_PASSWORD, UserFactory


PEAMA_SUB = "RANDOM_STRING"
PEAMA_ADDITIONAL_DATA = {
    "structureTravail": 59194,
    "siteTravail": "DRHAUTS-DE-FRANCE/PFPPLATEFORMESERVICESADISTANCE(HDF0262005733)",
}


def generate_peama_data(nonce):
    key_kid = "random_kid"

    key = jwk.JWK.generate(kty="RSA", size=2048, alg="RS256", use="sig", kid=key_kid)
    private_key = key.export_to_pem(private_key=True, password=None)
    jwk_key = key.export_public(as_dict=True)

    user_info = {
        "given_name": "Michel",
        "family_name": "AUDIARD",
        "email": "michel@pole-emploi.fr",
        "sub": PEAMA_SUB,
    }

    id_token_data = user_info | {"nonce": nonce} | PEAMA_ADDITIONAL_DATA
    id_token = jwt.encode(payload=id_token_data, key=private_key, algorithm="RS256", headers={"kid": key_kid})

    access_token = {
        "access_token": "00000000-0000-0000-0000-000000000000",
        "refresh_token": "11111111-1111-1111-1111-111111111111",
        "scope": settings.PEAMA_SCOPES,
        "id_token": id_token,
        "token_type": "Bearer",
        "expires_in": 1499,
    }

    jwks = {"keys": [jwk_key]}

    return SimpleNamespace(
        access_token=access_token,
        user_info=user_info,
        jwks=jwks,
    )


def mock_peama_oauth_dance(client, requests_mock, auth_complete_url):
    assert auth_complete_url.startswith(settings.PEAMA_AUTH_ENDPOINT)
    state = get_url_params(auth_complete_url)["state"]
    nonce = get_url_params(auth_complete_url)["nonce"]
    callback_url = reverse("oidc_federation:peama:callback")

    peama_data = generate_peama_data(nonce)
    requests_mock.post(settings.PEAMA_TOKEN_ENDPOINT, json=peama_data.access_token)
    requests_mock.get(settings.PEAMA_USER_ENDPOINT, json=peama_data.user_info)
    requests_mock.get(settings.PEAMA_JWKS_ENDPOINT, json=peama_data.jwks)

    return client.get(callback_url, data={"code": "123", "state": state}), peama_data


class TestFederation:
    def test_init_view(self, client):
        auth_url = reverse("oidc_federation:peama:init")
        response = client.get(auth_url)
        assert response.url.startswith(settings.PEAMA_AUTH_ENDPOINT)

        params = get_url_params(response.url)
        assert params["response_type"] == "code"
        assert params["scope"] == settings.PEAMA_SCOPES
        assert params["client_id"] == settings.PEAMA_CLIENT_ID
        assert params["redirect_uri"] == "http://testserver/federation/peama/callback/"
        assert params["realm"] == "/agent"
        assert list(params.keys()) == [
            "response_type",
            "scope",
            "client_id",
            "redirect_uri",
            "state",
            "realm",
            "nonce",
        ]

    def test_new_user(self, client, requests_mock, caplog):
        response = client.get(reverse("oidc_federation:peama:init"))
        response, peama_data = mock_peama_oauth_dance(client, requests_mock, response.url)
        user = User.objects.get()
        assert not user.email_addresses.exists()
        assert user.email == peama_data.user_info["email"]
        assert user.first_name == peama_data.user_info["given_name"]
        assert user.last_name == peama_data.user_info["family_name"]
        assert user.federation_sub == peama_data.user_info["sub"]
        assert user.federation == Federation.PEAMA
        assert user.federation_data == {
            "structure_pe": PEAMA_ADDITIONAL_DATA["structureTravail"],
            "site_pe": PEAMA_ADDITIONAL_DATA["siteTravail"],
        }
        assertRedirects(response, reverse("accounts:accept_terms"))
        assertRecords(
            caplog,
            [
                (
                    "inclusion_connect.auth.oidc_federation",
                    logging.INFO,
                    {"email": user.email, "user": user.pk, "event": "register", "federation": Federation.PEAMA},
                )
            ],
        )

    def test_update_existing_pe_user(self, client, requests_mock, caplog):
        user = UserFactory(
            email="old_email",
            first_name="old_first_name",
            last_name="old_last_name",
            federation_sub=PEAMA_SUB,
            federation=Federation.PEAMA,
        )
        response = client.get(reverse("oidc_federation:peama:init"))
        response, peama_data = mock_peama_oauth_dance(client, requests_mock, response.url)
        user = User.objects.get()
        assert user.email == peama_data.user_info["email"]
        assert user.first_name == peama_data.user_info["given_name"]
        assert user.last_name == peama_data.user_info["family_name"]
        assert user.federation_sub == peama_data.user_info["sub"]
        assert user.federation == Federation.PEAMA
        assert user.federation_data == {
            "site_pe": PEAMA_ADDITIONAL_DATA["siteTravail"],
            "structure_pe": PEAMA_ADDITIONAL_DATA["structureTravail"],
        }
        assertRedirects(response, reverse("accounts:edit_user_info"))
        assertRecords(
            caplog,
            [
                (
                    "inclusion_connect.auth.oidc_federation",
                    logging.INFO,
                    {"email": user.email, "user": user.pk, "event": "login", "federation": Federation.PEAMA},
                ),
                (
                    "inclusion_connect.auth.oidc_federation",
                    logging.INFO,
                    {
                        "event": "edit_user_info",
                        "user": user.pk,
                        "old_first_name": "old_first_name",
                        "new_first_name": peama_data.user_info["given_name"],
                        "old_last_name": "old_last_name",
                        "new_last_name": peama_data.user_info["family_name"],
                        "old_email": "old_email",
                        "new_email": peama_data.user_info["email"],
                        "old_federation_data": None,
                        "new_federation_data": user.federation_data,
                        "old_federation_id_token_hint": None,
                        "new_federation_id_token_hint": peama_data.access_token["id_token"],
                    },
                ),
            ],
        )

    def test_convert_existing_ic_user(self, client, requests_mock, caplog):
        UserFactory(
            email="michel@pole-emploi.fr",
            first_name="old_first_name",
            last_name="old_last_name",
        )
        response = client.get(reverse("oidc_federation:peama:init"))
        response, peama_data = mock_peama_oauth_dance(client, requests_mock, response.url)
        user = User.objects.get()
        assert user.email == peama_data.user_info["email"]
        assert user.first_name == peama_data.user_info["given_name"]
        assert user.last_name == peama_data.user_info["family_name"]
        assert user.federation_sub == peama_data.user_info["sub"]
        assert user.federation == Federation.PEAMA
        assert user.federation_data == {
            "structure_pe": PEAMA_ADDITIONAL_DATA["structureTravail"],
            "site_pe": PEAMA_ADDITIONAL_DATA["siteTravail"],
        }
        assertRedirects(response, reverse("accounts:edit_user_info"))
        assertRecords(
            caplog,
            [
                (
                    "inclusion_connect.auth.oidc_federation",
                    logging.INFO,
                    {"email": user.email, "user": user.pk, "event": "login", "federation": Federation.PEAMA},
                ),
                (
                    "inclusion_connect.auth.oidc_federation",
                    logging.INFO,
                    {
                        "event": "edit_user_info",
                        "user": user.pk,
                        "old_first_name": "old_first_name",
                        "new_first_name": peama_data.user_info["given_name"],
                        "old_last_name": "old_last_name",
                        "new_last_name": peama_data.user_info["family_name"],
                        "old_federation_sub": None,
                        "new_federation_sub": peama_data.user_info["sub"],
                        "old_federation": None,
                        "new_federation": Federation.PEAMA,
                        "old_federation_data": None,
                        "new_federation_data": user.federation_data,
                        "old_federation_id_token_hint": None,
                        "new_federation_id_token_hint": peama_data.access_token["id_token"],
                    },
                ),
            ],
        )

    def test_block_user_from_another_federation(self, client, requests_mock, caplog):
        user = UserFactory(
            email="michel@pole-emploi.fr",
            first_name="old_first_name",
            last_name="old_last_name",
            federation="other",
            federation_sub="sub",
        )
        response = client.get(reverse("oidc_federation:peama:init"))
        response, peama_data = mock_peama_oauth_dance(client, requests_mock, response.url)
        assertRedirects(response, reverse("accounts:login"))
        assert get_user(client).is_authenticated is False
        user = User.objects.get()
        assert user.email == peama_data.user_info["email"]
        assert user.first_name != peama_data.user_info["given_name"]
        assert user.last_name != peama_data.user_info["family_name"]
        assert user.federation_sub != peama_data.user_info["sub"]
        assert user.federation == "other"
        assert user.federation_data is None
        assertRecords(
            caplog,
            [
                (
                    "mozilla_django_oidc.auth",
                    logging.WARNING,
                    "failed to get or create user: email=michel@pole-emploi.fr from federation=peama is already used "
                    "by other",
                ),
                (
                    "inclusion_connect.auth.oidc_federation",
                    logging.INFO,
                    {"email": user.email, "user": user.pk, "event": "login_error", "federation": Federation.PEAMA},
                ),
            ],
        )

    def test_block_user_from_another_federation_same_sub(self, client, requests_mock, caplog):
        # This shouldn"t happen often since subs should be long random keys, but we never know
        UserFactory(
            email="michel@pole-emploi.fr",
            first_name="old_first_name",
            last_name="old_last_name",
            federation="other",
            federation_sub=PEAMA_SUB,
        )
        response = client.get(reverse("oidc_federation:peama:init"))
        response, peama_data = mock_peama_oauth_dance(client, requests_mock, response.url)
        assertRedirects(response, reverse("accounts:login"))
        assert get_user(client).is_authenticated is False
        user = User.objects.get()
        assert user.email == peama_data.user_info["email"]
        assert user.first_name != peama_data.user_info["given_name"]
        assert user.last_name != peama_data.user_info["family_name"]
        assert user.federation_sub == peama_data.user_info["sub"]
        assert user.federation == "other"
        assert user.federation_data is None
        assertRecords(
            caplog,
            [
                (
                    "mozilla_django_oidc.auth",
                    logging.WARNING,
                    "failed to get or create user: email=michel@pole-emploi.fr from federation=peama is already used "
                    "by other",
                ),
                (
                    "inclusion_connect.auth.oidc_federation",
                    logging.INFO,
                    {"email": user.email, "user": user.pk, "event": "login_error", "federation": Federation.PEAMA},
                ),
            ],
        )


@override_settings(PEAMA_ENABLED=None, PEAMA_CLIENT_ID=None, PEAMA_JWKS_ENDPOINT=None)
def test_dont_crash_if_not_configured(client):
    user = UserFactory()
    url = reverse("accounts:login")
    response = client.get(url)
    assertContains(response, "Connexion")
    assertContains(response, "Adresse e-mail")  # Ask for email, not username
    assertContains(response, reverse("accounts:register"))  # Link to register page

    response = client.post(url, data={"email": user.email, "password": DEFAULT_PASSWORD})
    assertRedirects(response, reverse("accounts:edit_user_info"), fetch_redirect_response=False)
