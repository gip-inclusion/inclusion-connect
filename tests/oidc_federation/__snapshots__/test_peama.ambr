# serializer version: 1
# name: test_updated_email_already_exists
  '''
  <main class="s-main" id="main" role="main">
  
              <section class="s-section pt-lg-3">
                  <div class="container">
                      
                      <div class="row g-0 bg-white">
                          
                              
      <div class="col-12 col-lg-6 bg-lg-white">
          <div class="bg-white p-3 py-lg-7 px-lg-8">
              <div class="container container-messages">
                  
                      
  
  
                  
              </div>
  
              
  
      <h1 class="h2">Votre adresse e-mail ne peut pas être mise à jour</h1>
  
      <div>
          <p>
              Votre fournisseur d'identité (Pôle emploi) nous indique que vous avez changé d’adresse e-mail.
              Cependant cette adresse e-mail existe déjà sur Inclusion Connect et ne peut pas être réutilisée.
          </p>
          <ul>
              <li>ancienne adresse (Inclusion Connect): old_email</li>
              <li>nouvelle adresse (transmise par Pôle emploi): michel@pole-emploi.fr</li>
          </ul>
          <p>
              Pour corriger la situation, merci de nous signaler le problème à <a href="mailto: support@connect.inclusion.beta.gouv.fr">support@connect.inclusion.beta.gouv.fr</a>.
          </p>
      </div>
  
      <form class="js-prevent-multiple-submit" method="post">
          <input name="csrfmiddlewaretoken" type="hidden" value="NORMALIZED_CSRF_TOKEN"/>
          <div>
              <button class="btn btn-block btn-primary" type="submit">Continuer</button>
          </div>
      </form>
  
  
  
          </div>
      </div>
  
                              
      <div class="d-none d-lg-flex col-lg-6 p-lg-6 justify-content-center align-items-center bg-light">
          <img alt="" class="img-fluid" loading="lazy" src="/static/img/illustration-ic.svg"/>
      </div>
  
                          
                      </div>
                  </div>
              </section>
  
          </main>
  '''
# ---
